src/mame/sinclair/specnext.cpp

```diff
@@ -0,0 +1,2757 @@
+		m_port_1ffd_data = (m_port_1ffd_data & ~0x01) | BIT(nr_wr_dat, 2);
+
+		memory_change(0x8e, nr_wr_dat);
+		LOGMEM("8e: 1f=%x 7f=%x df=%x\n", m_port_1ffd_data, m_port_7ffd_data, m_port_dffd_data);
+		break;
+	case 0x8f:
+		m_regs.nr_8f_mapping_mode = BIT(nr_wr_dat, 0, 2);
+		memory_change(0x8f, nr_wr_dat);
+		break;
+	case 0x90:
+		m_regs.nr_90_pi_gpio_o_en = (BIT(nr_wr_dat, 2, 6) << 2) | 0b00;
+		break;
+	case 0x91:
+		m_regs.nr_91_pi_gpio_o_en = nr_wr_dat;
+		break;
+	case 0x92:
+		m_regs.nr_92_pi_gpio_o_en = nr_wr_dat;
+		break;
+	case 0x93:
+		m_regs.nr_93_pi_gpio_o_en = BIT(nr_wr_dat, 0, 4);
+		break;
+	case 0x98:
+		m_regs.nr_98_pi_gpio_o = nr_wr_dat;
+		break;
+	case 0x99:
+		m_regs.nr_99_pi_gpio_o = nr_wr_dat;
+		break;
+	case 0x9a:
+		m_regs.nr_9a_pi_gpio_o = nr_wr_dat;
+		break;
+	case 0x9b:
+		m_regs.nr_9b_pi_gpio_o = BIT(nr_wr_dat, 0, 4);
+		break;
+	case 0xa0:
+		m_regs.nr_a0_pi_peripheral_en = nr_wr_dat;
+		break;
+	case 0xa2:
+		m_regs.nr_a2_pi_i2s_ctl = nr_wr_dat;
+		break;
+	case 0xa8:
+		m_regs.nr_a8_esp_gpio0_en = BIT(nr_wr_dat, 0);
+		break;
+	case 0xa9:
+		m_regs.nr_a9_esp_gpio0 = BIT(nr_wr_dat, 0);
+		break;
+	case 0xb8:
+		m_regs.nr_b8_divmmc_ep_0 = nr_wr_dat;
+		break;
+	case 0xb9:
+		m_regs.nr_b9_divmmc_ep_valid_0 = nr_wr_dat;
+		break;
+	case 0xba:
+		m_regs.nr_ba_divmmc_ep_timing_0 = nr_wr_dat;
+		break;
+	case 0xbb:
+		m_regs.nr_bb_divmmc_ep_1 = nr_wr_dat;
+		break;
+	case 0xc0:
+		m_regs.nr_c0_im2_vector = BIT(nr_wr_dat, 5, 3);
+		m_regs.nr_c0_stackless_nmi = BIT(nr_wr_dat, 3);
+		m_maincpu->nmi_stackless_w(m_regs.nr_c0_stackless_nmi);
+		m_regs.nr_c0_int_mode_pulse_0_im2_1 = BIT(nr_wr_dat, 0);
+		break;
+	case 0xc4:
+		m_regs.nr_c4_int_en_0_expbus = BIT(nr_wr_dat, 7);
+		m_regs.nr_22_line_interrupt_en = BIT(nr_wr_dat, 1);
+		break;
+	case 0xc6:
+		m_regs.nr_c6_int_en_2_654 = (BIT(nr_wr_dat, 6) << 2) | (BIT(nr_wr_dat, 5) << 1) | BIT(nr_wr_dat, 4);
+		m_regs.nr_c6_int_en_2_210 = (BIT(nr_wr_dat, 2) << 2) | (BIT(nr_wr_dat, 1) << 1) | BIT(nr_wr_dat, 0);
+		break;
+	case 0xcc:
+		m_regs.nr_cc_dma_int_en_0_7 = BIT(nr_wr_dat, 7);
+		m_regs.nr_cc_dma_int_en_0_10 = BIT(nr_wr_dat, 0, 2);
+		break;
+	case 0xcd:
+		m_regs.nr_cd_dma_int_en_1 = nr_wr_dat;
+		break;
+	case 0xce:
+		m_regs.nr_ce_dma_int_en_2_654 = BIT(nr_wr_dat, 4, 3);
+		m_regs.nr_ce_dma_int_en_2_210 = BIT(nr_wr_dat, 0, 3);
+		break;
+	case 0xd8:
+		m_regs.nr_d8_io_trap_fdc_en = BIT(nr_wr_dat, 0);
+		break;
+	default:
+		LOGWARN("wR: %X <- %x\n", nr_wr_reg, nr_wr_dat);
+		break;
+	}
+
+	m_sprites->mirror_index_w(0b111);
+	m_sprites->mirror_inc_w(0);
+}
+
+void specnext_state::nr_02_w(u8 nr_wr_dat)
+{
+	m_regs.nr_02_bus_reset = BIT(nr_wr_dat, 7);
+	m_regs.nr_02_iotrap = BIT(nr_wr_dat, 4);
+	m_regs.nr_02_generate_mf_nmi = BIT(nr_wr_dat, 3);
+	if (!m_regs.nr_02_generate_mf_nmi)
+		;
+	m_regs.nr_02_generate_divmmc_nmi = BIT(nr_wr_dat, 2);
+	if (!m_regs.nr_02_generate_divmmc_nmi)
+		;
+
+	if (BIT(nr_wr_dat, 1))
+		machine().schedule_hard_reset();
```

Comment:
“Hard reset” tears down the emulation session and starts the machine again.  Is that really what this should do?  It doesn’t really correspond to a real life thing.

Reply:
The next has two programmable resets hard and soft. That fits into MAME's ideology and I was thinking can be used this way. Maybe I'll reconsider and handle both somehow in softreset based on a condition.

Reply:
It fits MAME's terminology, but MAME's hard reset is by design something completely outside of the emulation, kind of like unplugging the machine entirely.  What are the soft and hard resets on the machine?  For instance, on the Apple //e the soft and hard resets both pull the CPU RESET line and the firmware does different things depending on which type it is.